### ISSUE TITLE
Can't compile: error: assignment of member ‘name’ in read-only object

### ISSUE THREAD

**Issue by UralZima** [zfsonlinux/zfs#7960]
Hello. I am using Gentoo and ZFS-master.
All was fine, until I decided to upgrade ZFS.
In gentoo for upgrade I need to merge 2 packages: zfs and zfs-kmod. 
Zfs was merged OK, but stopped working.

local ~ # zfs list
bad property list: invalid property 'name'

I thought it will work when latest kernel module will be merged. But it fails to compile.

  CC [M]  /var/tmp/portage/sys-fs/zfs-kmod-9999/work/zfs-kmod-9999/module/zfs/zfs_sysfs.o
  CC [M]  /var/tmp/portage/sys-fs/zfs-kmod-9999/work/zfs-kmod-9999/module/zfs/zfs_vfsops.o
/var/tmp/portage/sys-fs/zfs-kmod-9999/work/zfs-kmod-9999/module/zfs/zfs_sysfs.c: In function ‘zfs_kobj_add_attr’:
/var/tmp/portage/sys-fs/zfs-kmod-9999/work/zfs-kmod-9999/module/zfs/zfs_sysfs.c:151:38: error: assignment of member ‘name’ in read-only object
  zkobj->zko_attr_list[attr_num].name = attr_name;
                                      ^
/var/tmp/portage/sys-fs/zfs-kmod-9999/work/zfs-kmod-9999/module/zfs/zfs_sysfs.c:152:38: error: assignment of member ‘mode’ in read-only object
  zkobj->zko_attr_list[attr_num].mode = 0444;

Please help me fixing these issues.
Gentoo-Hardened, zfs-9999, zfs-kmod-9999, Linux-grsecurity-4.9.74, GCC 7.3.0

**Comment by drescherjm** [comment]
I can test under gentoo in a little while.

**Comment by GregorKopka** [comment]
> Hello. I am using Gentoo and ZFS-master.
> All was fine, until I decided to upgrade ZFS.
> In gentoo for upgrade I need to merge 2 packages: zfs and zfs-kmod.
> Zfs was merged OK, but stopped working.
> 
> local ~ # zfs list
> bad property list: invalid property 'name'

Could be a mismatch between kernel and userland ABI:
In case you upgraded from a way older version (=didn't `emerge zfs` for quite some time) you still might have the old userland binaries somewhere in your path, their location changed a while ago.
Similar problem (when you can only locate the new version in your path) could happen when the old kernel module is still loaded but the new version of the userland is used.

The compilation error could be caused in concert with grsecurity, see https://forums.grsecurity.net/viewtopic.php?f=3&t=3890 for a similar case (and possibly a solution).

**Comment by UralZima** [comment]
Thanks for replies. About zfs binary stopped working I know, it is because I compiled a new version of zfs and still use old version of zfs-kmod.

I am using this kernel for a long time and compilation of zfs worked well always. My last successful compilation was about 4-6 commits ago to the master. 

I read this link about grsecurity, but didn't found a solution. Also, I don't see any grsecurity errors at dmesg while compiling. 

So now I have new zfs and old zfs-kmod and I am afraid to reboot, I don't know how to compile the same version in gentoo or back-forward for some commits.

I think the problem is related to some new code in last commits, the kernel version or gcc version. 
I am using zfs (master) 1.5+ years at 4.9.24-grsec and about a year on 4.9.74-grsec (minipli), and compilation always was successful. 
Please help me resolve these issues. 
Glibc-2.26-r7, gcc-7.3.0-r5 hardened profile, binutils-2.31.1

**Comment by UralZima** [comment]
Hello. I tried to compile it on linux-4.9.131-dappersec, the latest grsec-updated kernel, and it also failed.
I have compiled it successfully on latest linux-4.18.14-gentoo. 
As I know, 4.9.x is a longterm, so it is kernel-version-related or grsecurity-related. But as I told you, 4-5 commits back it compiled and worked fine. 

Is it possible to fix master to work with 4.9.x grsecurity kernels? It was working for years before. Thanks.

**Comment by UralZima** [comment]
Hello. Anything on this? Because of some commit (I think it is related to https://github.com/zfsonlinux/zfs/commit/7a23c81342df05ace730bd303b4a73854dba43dd), I can't use dappersec kernel anymore. I also tried with 4.9.132-dappersec, and the error is the same.
It was always compatible with grsecurity patches, please fix it. Thanks so much

**Comment by UralZima** [comment]
Hello. Some update. I decided to go back to the commit it was working. I was looking at history of zfs_sysfs.c file https://github.com/zfsonlinux/zfs/commits/master/module/zfs/zfs_sysfs.c and found that it started to fail since commits from Sep 2. So it is not related to https://github.com/zfsonlinux/zfs/commit/7a23c81342df05ace730bd303b4a73854dba43dd
I compiled it successfully with a commit before that 
EGIT_COMMIT="bb91178e60df553071ce2e18b0067ef703f7b583"
I hope this information will help fix building on dappersec kernels. Thanks!

**Comment by behlendorf** [comment]
The proposed fix is here, http://paste.ubuntu.com/p/prrDP6Kv4z/  @vogelfreiheit can you please open a PR so the CI and test it and we can get it reviewed and integrated.

**Comment by UralZima** [comment]
I am ready to test it after PR is open and I am able to modify gentoo's ebuild to include it.
As I said, reverting to commit https://github.com/zfsonlinux/zfs/commit/7a23c81342df05ace730bd303b4a73854dba43dd not resolved the problem, I am able to use it on commit EGIT_COMMIT="bb91178e60df553071ce2e18b0067ef703f7b583", but the related file I think is zfs_sysfs.c

**Comment by vogelfreiheit** [comment]
Hi, I just saw this. I will let someone know so a PR can be opened, meanwhile I had some other fixes written for a few other spots that need testing.

**Comment by vogelfreiheit** [comment]
@behlendorf I was about to send some patches in, including this one, but apparently someone with oper rights in the channel was feeling like a primadonna and decided to issue a ban over hurt feelings related to a comment about the semantics of the word 'mostly'.... after PMS'ing around other folks. Did not seem like someone technically involved in the project, but he went ahead and did the same in -social. Therefore, odds are you did not get a message in the channel mentioning another issue that is present with the spl debug code (without further analysis it looks like misuse of snprintf(), leading to a single byte leak, but I haven't verified this proper). Holding on to the patches for now.

**Comment by kpande** [comment]
@vogelfreiheit  if you have patches, submit them. but name calling and hostility won't be tolerated even if you have patches to submit to the project. in your comment here, you decided to continue with hostility and name calling. it's a pattern you have exhibited for time, and a few users decided enough was enough. no one else complained about the loss. please behave professionally, and I will do the same.

**Comment by vogelfreiheit** [comment]
Eh, after reviewing that other issue where someone is calling for a Code of Conduct (which is a fairly dumb idea), I'm surprised to see that you assumed that person was me (perhaps the intention of the troll all along). Since this begs finally some proper attention, I will describe this issue (feel free to remove the comment later as you obviously are seemingly easily to have your feelings hurt regardless of whether the other person intended so or not...).

I'll just clarify a few things: I've never asked for help in the channel and treated someone "poorly". I responded ironically to ptx0's messages, after he spent the past few hours being dismissive or condescending to other people, obviously in some cases users with a very limited skill around ZFS, but that does not merit acting like an imbecile. None of them were insults or hostile, they simply pointed out that ptx0 was assuming things constantly and then replying with an air of arrogance and usually in a condescending tone, despite himself being far from a beacon of excellence as far as technical skill or coding go. Immediately after he proceeded to ban me from the channel, to which I responded in -social pointing out that his attitude and actions making him look puerile and petty regardless of the nature of the "offense" by someone else. This knee-jerk reaction came motivated more by petty personal sentiments (and sharing those with some others who, well, fit that profile) than any actually reasonable reasons.

Now, on the project: you are routinely accepting patches quite obviously without really doing your due diligence, ensuring proper QA, etc. Case in point, the one we have here. It's painful to see that you waste more time on petty nothings over actually making your code or merged patches look like something with a modicum technical quality. The control over whether contributors are just copy-pasting from other spots (*hint* licensing violations get people sued), if the patches actually conflict with major kernel features, or symbol exports, or..... is lacking all over place.

How about instead of making up half-truths about people offending your feelings, or your buddies' feelings, you go and properly benefit from the resources you so adamantly defend (and ask others to work with): How difficult was it to grab the patch I sent you from the link, test it, test it again, sign it off and merge it? You need to go around deleting comments and engaging in conversation with the other troll, but you are too busy to actually merge an out of band patch while you waste time over petty fights? Well, good luck with that. And you are surprised ZoL is not taken seriously anywhere but by the homelab/non-enterprise crowd. If you want to be a primadonna, at least have the cards in your sleeve to back it up: you can't justify it right now with the standard you are providing your users. And it runs in ring0, so go figure a better way to fend off all those users popping up on IRC asking about destroyed data because you merged a patch from a pull-request that would not stand a moment the LKML.

**Comment by bunder2015** [comment]
I am temporarily locking this thread.  If I could please ask everyone to calm down, I'm not about to pick sides in whatever conflict that is currently taking place here.  Brian has asked for someone to submit a PR to fix this issue.  Until that takes place, there is no reason for this issue to remain wide open given its current state.

**Comment by kpande** [comment]
> I'm surprised to see that you assumed that person was me (perhaps the intention of the troll all along).

I had no idea who it was.

> Since this begs finally some proper attention, I will describe this issue (feel free to remove the comment later as you obviously are seemingly easily to have your feelings hurt regardless of whether the other person intended so or not...).

that seems unnecessary.

> I'll just clarify a few things: I've never asked for help in the channel and treated someone "poorly".  I responded ironically to ptx0's messages, 

perhaps you were being unintentionally abrasive, as you were in the previous sentence where you imply I have easily hurt feelings. I'm sure you didn't mean to do that, right? then you can see how easily words can be misunderstood.

however, more than once have you come and ask for input (maybe not "help", exactly) and then were rude to the people who were asking you for more information. criticising how they were attempting to help you.

we are volunteers who want to help others. we don't want to deal with difficult people. bans aren't permanent, they can be removed.. sometimes bad things happen in life and we need to respond to them appropriately. people are known to have bad days.

coming onto the issue tracker and claiming to withhold your help because of an unrelated incident is not an appropriate response to being banned from an IRC channel.

> after he spent the past few hours being dismissive or condescending to other people, obviously in some cases users with a very limited skill around ZFS

I'd been curious who I'd abused so horribly that day and went back and found this:

```
[11:13:45] *** duairc <duairc!~shane@ana.rch.ist> has joined #zfsonlinux
[11:14:28] <duairc> So I set up a server using native ZFS encryption, and it's running great
[11:14:52] <duairc> The problem is, I lost the encryption key
[11:15:00] <duairc> The server is still running though
[11:15:04] <ptx0> hm
[11:15:08] <ptx0> is it unlocked now
[11:15:10] <duairc> Is there any way I can recover it from a running system?
[11:15:12] <duairc> Yes, is it
[11:15:14] <duairc> *it is
[11:15:23] <ptx0> i think you can change it, at least
[11:15:36] <duairc> I know with LUKS you have to enter the old key to change it, but maybe, I didn't actually try
[11:15:56] <ptx0> # zfs change-key rpool/crypt
[11:15:56] <ptx0> Enter new passphrase for 'rpool/crypt':
[11:16:03] <ptx0> not here, i guess
[11:25:44] <duairc> ptx0: Great, that saved me loads of work! Thank you.
[11:29:47] <ptx0> awesome
[11:29:57] <ptx0> backup :)
[11:48:07] <rjvb> Hmmm, helpful in some situations but not very secure, is it?
[11:51:50] <blackflow> totally not? I'm surprised it allows change without confirming with current phrase.
[12:15:55] <ptx0> yeah  well you could just bypass the check if you were so evil and had access
[12:16:00] <ptx0> i think you're worrying too much
[12:29:23] <ptx0> duairc: btw i keep hitting corruption with encryption
[12:29:39] <ptx0> recv'd dataset has IO errors
[12:29:50] <ptx0> the source seems fine though
[12:38:17] <duairc> ptx0: Hmmm, I haven't tried send/recv at all yet. Good to know though.

```

I couldn't find any instances of being condescending or dismissive of any new users, though I did help save someone from redoing hours of work and then warned them about a potential issue.

> but that does not merit acting like an imbecile. 

this is a part of a pattern of language from you that we have grown tired of. why do you think it's okay to call others imbeciles?

> None of them were insults or hostile, they simply pointed out that ptx0 was assuming things constantly and then replying with an air of arrogance and usually in a condescending tone, despite himself being far from a beacon of excellence as far as technical skill or coding go.

who are you talking about? who is "they"? I couldn't find this conversation. where did it happen?

> Immediately after he proceeded to ban me from the channel, to which I responded in -social pointing out that his attitude and actions making him look puerile and petty regardless of the nature of the "offense" by someone else. This knee-jerk reaction came motivated more by petty personal sentiments (and sharing those with some others who, well, fit that profile) than any actually reasonable reasons.

alright, it may have been a knee-jerk reaction. if you'd messaged me to work it out, things probably would have been Just Fine. interpersonal differences can be worked out, everyone in the project wants the same thing.

> Now, on the project: you are routinely accepting patches quite obviously without really doing your due diligence, ensuring proper QA, etc.

the ZFS test suite is  very new  and has many deficiencies. testing every possible scenario is, as you must be aware, the impossible task. if you've found problems with the process, raise them specifically as new issues so that they can be addressed. I haven't seen anything until just now, but you don't raise any specific issues or offer any solutions, which seems less than helpful.

> Case in point, the one we have here. It's painful to see that you waste more time 

but a lot of it has simply been spent waiting for a PR to be opened. if you're seeing it happen, why didn't you step up and submit one? is it because of the fact you've pointed out in the past that you consider our contribution process on github to be "for hipsters" and something you are essentially too good for?

> on petty nothings over actually making your code or merged patches look like something with a modicum of technical quality.

if you have any improvements you see you can make, please do so. otherwise you're just insulting the members of the project, again.

> The control over whether contributors are just copy-pasting from other spots (hint licensing violations get people sued), 

any license violations should be raised with the project and fixed. can you cite any specific cases? 

> if the patches actually conflict with major kernel features, or symbol exports, or..... is lacking all over the place.

I don't know what you're talking about. it should be easy to find cases where this is true if it's "all over the place".

> How about instead of making up half-truths about people offending your feelings, or your buddies' feelings, 

is it too much to ask you to not be rude to people when they give you advice? to not come on here and then insult everyone else's work - people who weren't even involved in your quarrel?

> How difficult was it to grab the patch I sent you from the link, test it, test it again, sign it off and merge it?

for what it's worth, I'm not the one who's responsible for merging patches - there is but one gateway for this project to have patches merged. I can review, test, ask for changes, but that's where it ends. Brian is the one who merges the changes.

I have to propose a corollary  to your question. how difficult is it to open a pull request? you mentioned back in early December you were going to open one.

> You need to go around deleting comments and engaging in conversation with the other troll

I deleted **your** comment here that only served to insult me and hold your PR hostage. if you don't want to submit the PR, don't - other users can do it when we find the time.

> but you are too busy to actually merge an out of band patch

it's a cooperative effort. you don't care to be a part of the project in the way we've required, and so your patch doesn't find inclusion. is this surprising to anyone?

> while you waste time over petty fights? Well, good luck with that.

observation: I don't know how long it took you to write your own comment here but it seems like you could have opened a PR in less time. 

> And you are surprised ZoL is not taken seriously anywhere but by the homelab/non-enterprise crowd. If you want to be a primadonna, 

this is another example of your pattern of disrespect. "You are surprised"? who? not taken seriously? that'd actually be news to... well.. most of us, I imagine.

> at least have the cards in your sleeve to back it up: you can't justify it right now with the standard you are providing your users. And it runs in ring0, so go figure a better way to fend off all those users popping up on IRC asking about destroyed data because you merged a patch from a pull-request that would not stand a moment in the LKML.

I'm not sure I quite follow. you've seen users reporting data loss with a released build? would that be v0.7.10 that we marked as On Fire shortly after release and that was never included in any distribution? maybe an `mmap()` related data loss back in 2013 from AIO write path errors? but that was never in a finished release.

it might actually be helpful to have a list of data issues that ZFS has had in the past, similar to CVEs.

**Comment by bunder2015** [comment]
@behlendorf hi, this patch appears to need some more work.

```
  CC [M]  /var/lib/buildbot/slaves/zfs/CentOS_6_x86_64__BUILD_/build/zfs/module/zfs/zfs_sysfs.o
cc1: warnings being treated as errors
/var/lib/buildbot/slaves/zfs/CentOS_6_x86_64__BUILD_/build/zfs/module/zfs/zfs_sysfs.c: In function ‘zfs_kobj_init_static_sysfsops’:
/var/lib/buildbot/slaves/zfs/CentOS_6_x86_64__BUILD_/build/zfs/module/zfs/zfs_sysfs.c:221: error: assignment discards qualifiers from pointer target type
/var/lib/buildbot/slaves/zfs/CentOS_6_x86_64__BUILD_/build/zfs/module/zfs/zfs_sysfs.c: At top level:
/var/lib/buildbot/slaves/zfs/CentOS_6_x86_64__BUILD_/build/zfs/module/zfs/zfs_sysfs.c:393: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘{’ token
/var/lib/buildbot/slaves/zfs/CentOS_6_x86_64__BUILD_/build/zfs/module/zfs/zfs_sysfs.c: In function ‘zfs_kernel_features_init’:
/var/lib/buildbot/slaves/zfs/CentOS_6_x86_64__BUILD_/build/zfs/module/zfs/zfs_sysfs.c:403: error: ‘zfs_kernel_features_sysfs_ops’ undeclared (first use in this function)
/var/lib/buildbot/slaves/zfs/CentOS_6_x86_64__BUILD_/build/zfs/module/zfs/zfs_sysfs.c:403: error: (Each undeclared identifier is reported only once
/var/lib/buildbot/slaves/zfs/CentOS_6_x86_64__BUILD_/build/zfs/module/zfs/zfs_sysfs.c:403: error: for each function it appears in.)
/var/lib/buildbot/slaves/zfs/CentOS_6_x86_64__BUILD_/build/zfs/module/zfs/zfs_sysfs.c: In function ‘zfs_sysfs_properties_init’:
/var/lib/buildbot/slaves/zfs/CentOS_6_x86_64__BUILD_/build/zfs/module/zfs/zfs_sysfs.c:583: error: assignment discards qualifiers from pointer target type
/var/lib/buildbot/slaves/zfs/CentOS_6_x86_64__BUILD_/build/zfs/module/zfs/zfs_sysfs.c:593: error: assignment discards qualifiers from pointer target type
make[5]: *** [/var/lib/buildbot/slaves/zfs/CentOS_6_x86_64__BUILD_/build/zfs/module/zfs/zfs_sysfs.o] Error 1
```

**Comment by ahrens** [comment]
The discussion in this thread (now hidden) has not been productive.  In the future, please speak respectfully to fellow contributors.  @behlendorf and I are working on putting a process in place for handling inappropriate behavior.  In the mean time, please feel free to contact either Brian or I privately if you have concerns about the behavior of ZFSonLinux community members.
